### PR TITLE
feature/use RemainingItemCount feature gate

### PR DESCRIPTION
**Pull request by feeltimeQ**
## issue: #81

if RemainingItemCount=true
wecom-temp-b80f1520bbcabf1829850fe674995cb8

if RemainingItemCount=false
wecom-temp-d653afaa4db30d7ddde498ad59d7f71a

### CODE REVIEW

**Review by cleverhu — COMMENTED** [review]
When the user set the WithRemainingCount to false, and does not want to return the remaining count in order to speed up the search efficiency, it will waste redundant resources. Therefore, it is necessary to judge whether the WithRemainingCount is nil.

File: pkg/kubeapiserver/resourcerest/storage.go [hunk 1]
```diff
@@ -70,6 +72,7 @@ func (s *RESTStorage) List(ctx context.Context, _ *metainternalversion.ListOptio
 }
 
 func (s *RESTStorage) list(ctx context.Context, options *internal.ListOptions) (runtime.Object, error) {
+	defaultRemainingItemCount := utilfeature.DefaultFeatureGate.Enabled(defaultfeatures.RemainingItemCount)
```
Comment (cleverhu) [inline pkg/kubeapiserver/resourcerest/storage.go:75]: This value is required when remaincount is nil, you don't need to judge at the beginning every time.

Comment (cleverhu) [reply pkg/kubeapiserver/resourcerest/storage.go:75]: The variable defaultremainingitemcount means whether it is enabled. Using this name will be a little tempting.

**Comment by cleverhu** [comment]
It is suggested that query should also be supported. There is still some extra work to be done.

File: pkg/kubeapiserver/resourcerest/storage.go [hunk 2]
```diff
@@ -10,8 +10,10 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	genericrequest "k8s.io/apiserver/pkg/endpoints/request"
+	defaultfeatures "k8s.io/apiserver/pkg/features"
```
Comment (Iceber) [inline pkg/kubeapiserver/resourcerest/storage.go:13]: The usual alias is `genericfeatures`

File: pkg/kubeapiserver/resourcerest/storage.go [hunk 3]
```diff
@@ -87,11 +89,12 @@ func (s *RESTStorage) list(ctx context.Context, options *internal.ListOptions) (
 		return nil, apierrors.NewBadRequest("If searching by owner uid or name, then the cluster must be specified")
 	}
 
-	/*
-		TODO(iceber): add feature gates to set default option
-		if options.WithRemainingCount == nil {
+	if options.WithRemainingCount == nil {
+		if utilfeature.DefaultFeatureGate.Enabled(defaultfeatures.RemainingItemCount) != false {
```
Comment (Iceber) [inline pkg/kubeapiserver/resourcerest/storage.go:93]: Suggestion:
```go
if enabled := utilfeature.DefaultFeatureGate.Enabled(defaultfeatures.RemainingItemCount); enabled {
    options.WithRemainingCount = &enabled
}
```

**Review by Iceber — COMMENTED** [review]
Commit looks like a branch name that can be changed to `feature: use RemainingItemCount feature gate`

DCO Check is failed.
> There is one commit incorrectly signed off. This means that the author of this commit failed to include a Signed-off-by line in the commit message.
> 
> To avoid having PRs blocked in the future, always include Signed-off-by: Author Name <authoremail@example.com> in every commit message. You can also do this automatically by using the -s flag (i.e., git commit -s).
> 
> Here is how to fix the problem so that this code can be merged.

**Review by Iceber — APPROVED** [review]
LGTM

**Comment by Iceber** [comment]
@calvin0327 @cleverhu The default value for feature gateway `RemainingItemCount` is true.
We can't change this default, so it seems we need to actively set these in the charts and deploy yaml. `--feature-gates=RemainingItemCount=true

**Comment by Iceber** [comment]
The default value for feature gateway RemainingItemCount is true.

We can't change this default, so it seems we need to actively set it in the charts and  apiserver deploy yaml. `--feature-gates=RemainingItemCount=false`.
https://github.com/clusterpedia-io/clusterpedia/blob/e38f39f03d2483ebbfc5b2c96a13e6b71a0f835f/deploy/clusterpedia_apiserver_deployment.yaml#L36-L44
What do you think? @calvin0327 @cleverhu

**Pull request merged**